src/server/context.ts

```diff
@@ -536,6 +613,33 @@ const DEFAULT_ERROR: ErrorPage = {
   csp: false,
 };
 
+/**
+ * Return a list of middlewares that needs to be applied for request url
+ * @param url the request url
+ * @param middlewares Array of middlewares handlers and their routes as path-to-regexp style
+ * @returns
+ */
+export function selectMiddlewares(url: string, middlewares: MiddlewareRoute[]) {
+  const selectedMw: [number, Middleware][] = [];
+  const reqURL = new URL(url);
+
+  for (const { route, handler } of middlewares) {
+    const regRoute = pathToRoute(route).slice(0, -"_middleware".length);
+    // check for each subpath of url if it is matching the middleware path
+    const pattern = new URLPattern({
+      pathname: `${regRoute}*`,
+    });
+    const res = pattern.exec(reqURL);
+    if (res) {
+      selectedMw.push([route.split("/").length, { handler }]);
+    }
+  }
+
+  // sort middlewares to apply by their layer level from the deepest layer to the root
+  const mws = selectedMw.sort((a, b) => b[0] - a[0]).map((mw) => mw[1]);
```

Comment:
The ordering of middlewares relative to each other can also be done ahead of time to prevent some overhead here, as the ordering is fixed.